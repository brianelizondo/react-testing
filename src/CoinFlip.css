.CoinFlip{
    font-family: Verdana, Geneva, Tahoma, sans-serif;
    margin-bottom: 100px;
}
.CoinFlip-button{
    padding-top: 30px;
    padding-bottom: 20px;
}
.CoinFlip-button button{ 
    border-radius: 5px;
    font-size: 11px;
    text-transform: uppercase;
    padding: 5px 20px;
}
.CoinFlip-button button:hover{ 
    background-color: gray;
    color: white;
}
.CoinFlip-track{
    font-size: 10px;
}