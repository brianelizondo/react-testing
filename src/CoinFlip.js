import React, { useState } from "react";
import "./CoinFlip.css";
import coin_head from "./coin_head.jpg";
import coin_tail from "./coin_tail.jpg";
import Coin from "./Coin";

function CoinFlip(props) {
    const [coinData, setCoinData] = useState({ src: "", face: "", isShow: false });
    const [flips, setFlips] = useState(0);
    const [flipsHead, setFlipsHead] = useState(0);
    const [flipsTail, setFlipsTail] = useState(0);

    const newFlip = () => {
        const randomIDX = Math.floor(Math.random() * props.coins.length);
        const newCoin = props.coins[randomIDX];
        setCoinData(newCoin);

        setFlips(flips + 1);
        if(newCoin.face === "head"){
            setFlipsHead(flipsHead + 1);
        }else{
            setFlipsTail(flipsTail + 1);
        }
    }

    return (
        <div className="CoinFlip">
            <h1>Let's flip a coin</h1>
            <Coin src={ coinData.src } face={ coinData.face } isShow={ coinData.isShow } />
            <div className="CoinFlip-button"><button onClick={ newFlip } data-testid="flip-me">flip meeee</button></div>
            <div className="CoinFlip-track">
                Out of <b>{ flips }</b> flips, 
                there haven been <b>{ flipsHead } heads </b> 
                and <b>{ flipsTail } tails</b>
            </div>
        </div>
    );
}

CoinFlip.defaultProps = {
    coins: [
        {
            src: coin_head,
            face: "head",
            isShow: true
        },
        {
            src: coin_tail,
            face: "tail",
            isShow: true
        }
    ]
};

export default CoinFlip;