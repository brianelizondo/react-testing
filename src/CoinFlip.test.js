import React from "react";
import { render, fireEvent } from "@testing-library/react";
import CoinFlip from "./CoinFlip";

// Smoke Tests
it("'CoinFlip' component renders without crashing", function() {
    render(<CoinFlip />);
});

// Snapshot Tests
it("'CoinFlip' matches snapshot", function() {
    const {asFragment} = render(<CoinFlip />);
    expect(asFragment()).toMatchSnapshot();
});

it("works before click on the flip button", function() {
    const { queryByTestId, queryByAltText } = render(<CoinFlip />);
  
    // expect not show a coin
    expect(queryByAltText("head")).not.toBeInTheDocument();
    expect(queryByAltText("tail")).not.toBeInTheDocument();
    expect(queryByTestId("flip-me")).toBeInTheDocument();
  
});

it("works when click on the flip button and show HEAD", function() {
    const { queryByTestId, queryByAltText } = render(<CoinFlip />);
  
    // expect not show a coin
    expect(queryByAltText("head")).not.toBeInTheDocument();
    expect(queryByAltText("tail")).not.toBeInTheDocument();
    expect(queryByTestId("flip-me")).toBeInTheDocument();
  
    // flip the coin button
    const flipButton = queryByTestId("flip-me");
    fireEvent.click(flipButton);
  
    // expect to show a coin
    expect(queryByAltText("head")).toBeInTheDocument();
    expect(queryByAltText("tail")).not.toBeInTheDocument();
});

it("works when click on the flip button and show TAIL", function() {
    const { queryByTestId, queryByAltText } = render(<CoinFlip />);
  
    // expect not show a coin
    expect(queryByAltText("head")).not.toBeInTheDocument();
    expect(queryByAltText("tail")).not.toBeInTheDocument();
    expect(queryByTestId("flip-me")).toBeInTheDocument();
  
    // flip the coin button
    const flipButton = queryByTestId("flip-me");
    fireEvent.click(flipButton);
  
    // expect to show a coin
    expect(queryByAltText("head")).not.toBeInTheDocument();
    expect(queryByAltText("tail")).toBeInTheDocument();
});