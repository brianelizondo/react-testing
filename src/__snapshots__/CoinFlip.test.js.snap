// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'CoinFlip' matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CoinFlip"
  >
    <h1>
      Let's flip a coin
    </h1>
    <div
      class="Coin"
    />
    <div
      class="CoinFlip-button"
    >
      <button
        data-testid="flip-me"
      >
        flip meeee
      </button>
    </div>
    <div
      class="CoinFlip-track"
    >
      Out of 
      <b>
        0
      </b>
       flips, there haven been 
      <b>
        0 heads 
      </b>
      and 
      <b>
        0 tails
      </b>
    </div>
  </div>
</DocumentFragment>
`;
